File: mobile/android/OpenU60/app/src/main/java/com/openu60/core/model/ClashModels.kt
package com.openu60.core.model

data class ClashStatus(
    val running: Boolean = false,
    val version: String = "",
    val mode: String = "",
    val controllerPort: Int = 7788,
    val ui: String = "",
    val mixedPort: Int = 0,
    val redirPort: Int = 0,
    val tproxyPort: Int = 0,
    val allowLan: Boolean = false,
    val tunEnabled: Boolean = false,
    val connections: Int = 0,
    val downloadTotal: Long = 0,
    val uploadTotal: Long = 0,
) {
    companion object {
        val empty = ClashStatus()
    }

    val modeLabel: String
        get() = when (mode.lowercase()) {
            "rule" -> "规则"
            "global" -> "全局"
            "direct" -> "直连"
            else -> "--"
        }
}

data class ClashSelectorGroup(
    val name: String = "",
    val current: String = "",
    val options: List<String> = emptyList(),
    val alive: Boolean = true,
) {
    companion object {
        val empty = ClashSelectorGroup()
    }
}

object ClashParser {
    fun offlineStatus(): ClashStatus = ClashStatus.empty

    fun parseStatus(
        version: Map<String, Any?>,
        configs: Map<String, Any?>,
        connections: Map<String, Any?>,
    ): ClashStatus {
        return ClashStatus(
            running = true,
            version = version["version"] as? String ?: "",
            mode = configs["mode"] as? String ?: "",
            controllerPort = 7788,
            ui = "zashboard",
            mixedPort = DeviceParser.asInt(configs["mixed-port"]) ?: 0,
            redirPort = DeviceParser.asInt(configs["redir-port"]) ?: 0,
            tproxyPort = DeviceParser.asInt(configs["tproxy-port"]) ?: 0,
            allowLan = DeviceParser.asBool(configs["allow-lan"]),
            tunEnabled = DeviceParser.asBool((configs["tun"] as? Map<*, *>)?.get("enable")),
            connections = (connections["connections"] as? List<*>)?.size ?: 0,
            downloadTotal = DeviceParser.asLong(connections["downloadTotal"]) ?: 0,
            uploadTotal = DeviceParser.asLong(connections["uploadTotal"]) ?: 0,
        )
    }

    fun parseSelectors(data: Map<String, Any?>): List<ClashSelectorGroup> {
        val proxies = data["proxies"] as? Map<*, *> ?: return emptyList()
        return proxies.mapNotNull { (rawName, rawValue) ->
            val name = rawName?.toString() ?: return@mapNotNull null
            val item = rawValue as? Map<*, *> ?: return@mapNotNull null
            if ((item["type"] as? String) != "Selector") return@mapNotNull null
            val options = (item["all"] as? List<*>)?.mapNotNull { it?.toString() } ?: emptyList()
            if (options.isEmpty()) return@mapNotNull null
            ClashSelectorGroup(
                name = name,
                current = item["now"] as? String ?: "",
                options = options,
                alive = DeviceParser.asBool(item["alive"]),
            )
        }.sortedBy { it.name }
    }
}

File: mobile/ios/OpenU60/Core/Models/ClashModels.swift
import Foundation

struct ClashStatus: Equatable {
    var running: Bool
    var version: String
    var mode: String
    var controllerPort: Int
    var ui: String
    var mixedPort: Int
    var redirPort: Int
    var tproxyPort: Int
    var allowLan: Bool
    var tunEnabled: Bool
    var connections: Int
    var downloadTotal: UInt64
    var uploadTotal: UInt64

    static let empty = ClashStatus(
        running: false,
        version: "",
        mode: "",
        controllerPort: 7788,
        ui: "",
        mixedPort: 0,
        redirPort: 0,
        tproxyPort: 0,
        allowLan: false,
        tunEnabled: false,
        connections: 0,
        downloadTotal: 0,
        uploadTotal: 0
    )

    var modeLabel: String {
        switch mode.lowercased() {
        case "rule": return "规则"
        case "global": return "全局"
        case "direct": return "直连"
        default: return "--"
        }
    }
}

struct ClashSelectorGroup: Equatable, Identifiable {
    let id: String
    var name: String
    var current: String
    var options: [String]
    var alive: Bool
}

enum ClashParser {
    static func offlineStatus() -> ClashStatus { .empty }

    static func parseStatus(_ version: [String: Any], _ configs: [String: Any], _ connections: [String: Any]) -> ClashStatus {
        ClashStatus(
            running: true,
            version: version["version"] as? String ?? "",
            mode: configs["mode"] as? String ?? "",
            controllerPort: 7788,
            ui: "zashboard",
            mixedPort: asInt(configs["mixed-port"]) ?? 0,
            redirPort: asInt(configs["redir-port"]) ?? 0,
            tproxyPort: asInt(configs["tproxy-port"]) ?? 0,
            allowLan: asBool(configs["allow-lan"]),
            tunEnabled: asBool((configs["tun"] as? [String: Any])?["enable"]),
            connections: (connections["connections"] as? [[String: Any]])?.count ?? 0,
            downloadTotal: asUInt64(connections["downloadTotal"]) ?? 0,
            uploadTotal: asUInt64(connections["uploadTotal"]) ?? 0
        )
    }

    static func parseSelectors(_ data: [String: Any]) -> [ClashSelectorGroup] {
        guard let proxies = data["proxies"] as? [String: [String: Any]] else { return [] }
        return proxies.compactMap { name, item in
            guard item["type"] as? String == "Selector" else { return nil }
            let options = item["all"] as? [String] ?? []
            guard !options.isEmpty else { return nil }
            return ClashSelectorGroup(
                id: name,
                name: name,
                current: item["now"] as? String ?? "",
                options: options,
                alive: asBool(item["alive"])
            )
        }
    }

    private static func asBool(_ value: Any?) -> Bool {
        if let str = value as? String {
            return str == "1" || str.lowercased() == "true" || str.lowercased() == "on"
        }
        if let num = value as? Int { return num != 0 }
        if let num = value as? NSNumber { return num.intValue != 0 }
        if let b = value as? Bool { return b }
        return false
    }

    private static func asInt(_ value: Any?) -> Int? {
        if let int = value as? Int { return int }
        if let number = value as? NSNumber { return number.intValue }
        if let string = value as? String { return Int(string) }
        return nil
    }

    private static func asUInt64(_ value: Any?) -> UInt64? {
        if let int = value as? UInt64 { return int }
        if let number = value as? NSNumber { return number.uint64Value }
        if let string = value as? String { return UInt64(string) }
        return nil
    }
}
